### ISSUE TITLE
Bunch acceleration for electron injection into PWFA stage

### ISSUE THREAD

**Issue by berceanu** [ComputationalRadiationPhysics/picongpu#2808]
Following discussions with @BeyondEspresso, I am trying to make make use of @steindev's bunch acceleration example  from [this branch](https://github.com/steindev/picongpu/tree/topic-2018_bunch-accel-ulf/share/picongpu/examples/BunchAcceleration). This will be useful in my case for setting up PWFA simulations.

However, compilation seems to fail with `PIConGPU 0.4.1`, see [pic-build log](https://gist.githubusercontent.com/berceanu/5b3f8eaf10d38ec8cdf2b441b15a982a/raw/6f5ef2c03a0e2b7c37b1fc94c302dbbdf7024ace/pic-build.out).

```
$ export VERBOSE=ON; pic-build
[...]
include/pmacc/../picongpu/traits/GetMargin.hpp(41): error: incomplete type is not allowed
          detected during:
            instantiation of class "picongpu::traits::GetMargin<Solver, SubSetName> [with Solver=picongpu::UsedParticlePusher, SubSetName=0U]" 
(48): here
            instantiation of class "picongpu::traits::GetLowerMargin<T_Type> [with T_Type=picongpu::UsedParticlePusher]" 
[...]
```

PS: If you guys think it would be useful, I would like to create a full example out of this and add it to the docs.

**Comment by sbastrakov** [comment]
Thanks for the bug report. Can't check this right now, but the reason might be `typename` was forgotten in the implementation [here](https://github.com/ComputationalRadiationPhysics/picongpu/blob/a144ee4bb14b5c4c58aacc290fe4dfb6f5ab88b6/include/picongpu/particles/pusher/particlePusherAcceleration.hpp#L50-L51) and in the next line.

**Comment by sbastrakov** [comment]
Now I see that the example sets the force, so the accelerating pusher is required. I guess you could try the latest dev, or the branch where the example is (or, if 0.4.1 is absolutely required, try to manually apply the #2731 changes that added the pusher).

**Comment by steindev** [comment]
Yes you should use the current `dev` branch of picongpu.  The acceleration pusher is not implemented in 0.4.1. (ax3l explained this in #2784)

**Comment by berceanu** [comment]
Indeed, that is explained in https://github.com/ComputationalRadiationPhysics/picongpu/pull/2784#issuecomment-435777343.

Sorry about this @steindev, I got confused. I understood there are two methods of electron injection, one based on the new pusher and another which doesn't require any code changes. I somehow thought [your branch](https://github.com/steindev/picongpu/tree/topic-2018_bunch-accel-ulf/) was the latter.
